### PR TITLE
[manifest] #3 add a module to abstract a manifest file

### PR DESCRIPTION
**Pull request by PengTian0**
Add a module to abstract a manifest file

The proposal of manifest build in rackhd release  is [here](https://groups.google.com/forum/#!topic/rackhd/v3LpRuQJvQg).

This PR depends on https://github.com/RackHD/on-tools/pull/25 and https://github.com/RackHD/on-tools/pull/26

### CODE REVIEW

**Comment by PengTian0** [comment]
@panpan0000

**Comment by PengTian0** [comment]
@RackHD/corecommitters

**Comment by panpan0000** [comment]
will you add your pretty manifest README.md ? in other PR?

**Comment by PengTian0** [comment]
@panpan0000 Yes!
I will add documents for manifest after the manifest repository is created.

File: manifest-build-tools/lib/manifest.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    def validate_downstream_jobs(downstream_jobs):
+        """
+        validate whether the entry 'downstream-jobs' contains useful information
+        return: True if the entry is valid,
+                False if the entry is unusable
+                and message including omission imformation
+        """
+        result = True
+        message = []
+        #for job in self._manifest['downstream-jobs']:
```
Comment (panpan0000) [inline manifest-build-tools/lib/manifest.py]: abandoned line ?

File: manifest-build-tools/lib/manifest.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        """
+
+        self._build_name = None
+        self._build_requirements = None
+        self._repositories = []
+        self._downstream_jobs = []
+        self._file_path = file_path
+        self._name = file_path.split('/')[-1]
+        self._manifest = None
+        self._changed = False
```
Comment (panpan0000) [inline manifest-build-tools/lib/manifest.py]: the self._changed is not set inside this class ?

Comment (PengTian0) [inline manifest-build-tools/lib/manifest.py]: Oh, yes.
The member is used in updating manifest.
I will submit another commit to remove it.

File: manifest-build-tools/lib/manifest.py (hunk 1 above)
Comment (PengTian0) [inline manifest-build-tools/lib/manifest.py]: Yes, this line is abandoned.
I will remove it.

File: manifest-build-tools/lib/manifest.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        self.gitbit = GitBit(verbose=True)
+
+        if git_credentials:
+            self._git_credentials = git_credentials
+            self.setup_gitbit()
+
+        self.read_manifest_file(self._file_path)
+        self.parse_manifest()
+
+    def set_git_credentials(self, git_credentials):
```
Comment (lanchongyizu) [inline manifest-build-tools/lib/manifest.py]: Does update_git_credentials make more sense?

File: manifest-build-tools/lib/manifest.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        """
+        __build_name - Refer to the field "build-name" in manifest file.
+        __repositories - Refer to the field "repositories" in manifest file.
+        __downstream_jobs - Refer to the field "downstream-jobs" in manifest file.
+        __file_path - The file path of the manifest file
+        __name - The file name of the manifest file
+        __manifest -The content of the manifest file
+        __git_credentials  - URL, credentials pair for the access to github repos
+        gitbit - Class instance of gitbit
+        """
```
Comment (lanchongyizu) [inline manifest-build-tools/lib/manifest.py]: The comments format is incorrect.

File: manifest-build-tools/lib/manifest.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        self._build_name = None
+        self._build_requirements = None
+        self._repositories = []
+        self._downstream_jobs = []
+        self._file_path = file_path
+        self._name = file_path.split('/')[-1]
+        self._manifest = None
+
+        self._git_credentials = None
+        self.gitbit = GitBit(verbose=True)
```
Comment (lanchongyizu) [inline manifest-build-tools/lib/manifest.py]: Is gitbit private?

File: manifest-build-tools/lib/manifest.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    def get_file_path(self):
+        return self._file_path
+
+    def get_build_name(self):
+        return self._build_name
+
+    def get_build_requirements(self):
+        return self._build_requirements
+
+
```
Comment (lanchongyizu) [inline manifest-build-tools/lib/manifest.py]: Delete one of the blank lines.

File: manifest-build-tools/lib/manifest.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    def setup_gitbit(self):
+        """
+        Set gitbit credentials.
+        :return:
```
Comment (lanchongyizu) [inline manifest-build-tools/lib/manifest.py]: :return: None

File: manifest-build-tools/lib/manifest.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+            r, m = self.validate_repositories(self._repositories)
+            if not r:
+                result = False
+                message.extend(m)
+        #downstream-jobs is required
+        if 'downstream-jobs' not in self._manifest:
+            result = False
+            message.append("No downstream-jobs in manifest file")
+        else:
+            r, m = self.validate_downstream_jobs(self._downstream_jobs)
```
Comment (lanchongyizu) [inline manifest-build-tools/lib/manifest.py]: validate_downstream_jobs is static method

File: manifest-build-tools/lib/manifest.py [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+
+        #repositories is required
+        if 'repositories' not in self._manifest:
+            result = False
+            message.append("No repositories in manifest file")
+        else:
+            r, m = self.validate_repositories(self._repositories)
+            if not r:
+                result = False
+                message.extend(m)
```
Comment (lanchongyizu) [inline manifest-build-tools/lib/manifest.py]: messages is more meaningful

File: manifest-build-tools/lib/manifest.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+        if 'build-name' not in self._manifest:
+            result = False
+            message.append("No build-name in manifest file")
+
+        #repositories is required
+        if 'repositories' not in self._manifest:
+            result = False
+            message.append("No repositories in manifest file")
+        else:
+            r, m = self.validate_repositories(self._repositories)
```
Comment (lanchongyizu) [inline manifest-build-tools/lib/manifest.py]: Change r and m to meaningful name.

File: manifest-build-tools/lib/manifest.py [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+        """
+        validate whether the entry 'downstream-jobs' contains useful information
+        return: True if the entry is valid,
+                False if the entry is unusable
+                and message including omission imformation
+        """
+        result = True
+        message = []
+        for job in downstream_jobs:
+            valid = True
```
Comment (lanchongyizu) [inline manifest-build-tools/lib/manifest.py]: Why use result and valid? valid is better than result.
The same problem exists in validate_* methods.

File: manifest-build-tools/lib/manifest.py (hunk 6 above)
Comment (PengTian0) [inline manifest-build-tools/lib/manifest.py]: Done!

File: manifest-build-tools/lib/manifest.py (hunk 7 above)
Comment (PengTian0) [inline manifest-build-tools/lib/manifest.py]: Done!

File: manifest-build-tools/lib/manifest.py (hunk 11 above)
Comment (PengTian0) [inline manifest-build-tools/lib/manifest.py]: valid means a repository is validated or not.
result is the result of the function.

File: manifest-build-tools/lib/manifest.py (hunk 9 above)
Comment (PengTian0) [inline manifest-build-tools/lib/manifest.py]: Yes.
I will submit a commit to change it.

File: manifest-build-tools/lib/manifest.py (hunk 8 above)
Comment (PengTian0) [inline manifest-build-tools/lib/manifest.py]: Oh, yes.
I will fix it!

**Comment by lanchongyizu** [comment]
:+1:

**Pull request merged**